src/libopensc/card-piv.c

```diff
@@ -3760,12 +3760,15 @@ piv_pin_cmd(sc_card_t *card, struct sc_pin_cmd_data *data, int *tries_left)
 
 static int piv_logout(sc_card_t *card)
 {
-	int r = SC_ERROR_NOT_SUPPORTED; /* TODO Some PIV cards may support a logout */
-	/* piv_private_data_t * priv = PIV_DATA(card); */
+	int r, pin_ref;
 
 	LOG_FUNC_CALLED(card->ctx);
 
-	/* TODO 800-73-3 does not define a logout, 800-73-4 does */
+	r = piv_get_pin_preference(card, &pin_ref);
+	LOG_TEST_RET(card->ctx, r, "Failure retrieving PIN reference");
```

Comment:
should the failure here cause the `sc_reset()` on cards not supporting the -4 specs?

Reply:
Doing a reset for PIV is generally not a good idea, because a PIV *applet* may then not be selected anymore. I realised that @dengert has an even more relaxed handling of errors (i.e. [returning SUCCESS if logout fails](https://github.com/OpenSC/OpenSC/blob/c853dd4df834e461500b015eabfd1ea49a44b757/src/libopensc/card-piv.c#L6147-L6148)). I think I would go with Doug's implementation, but return the error to the upper layer if it fails. For example, in CTK we can handle this error by still treating the card as *sensitive* (i.e. to be separated from other processes).

Reply:
For the record the PIV driver at the start of **every** transaction will either do GET_DATA for the DISCOVERY object and/or a  SELECT AID. (If DISCOVERY is not found or useless it will do a SELECT AID.)  Some "PIV like" cards would lose login in state even if SELECT AID for PIV was used!

From day one in 2004, the goal was to keep the PIV login state active so it could be used by multiple applications especially with PIN PAD reader  even after an application ends. With cards with multiple applet, this is a problem, as NIST 800-73 specially says if an other applet that is on the card is selected, the PIV applet login state is reset. Some "PIV like" cards will also reset the login state if a SELECT_AID is done even if the card does have that applet. So the goal is not possible with some cards or with some non-OpenSC applications. 

NIST only defined the logout in 800-73-4 But does not require it be used.

The use of `opensc.conf` `disconnect_action = leave;` is really the way a user can say to leave the card in a logged in state, and not do a reset or logout when application is done. `disconnect_action = reset;` 

IMHO the decision to do a reset or logout should only be done when `disconnect_action = reset;`  or `disconnect_action = unpower;`

Reply:
If I understood correctly, you want to use the logout command from 800-73-4 and if that fails we should do a reset of the card? That would be easy to do and we would just have to reselect the PIV AID to make sure the applet is active.

Reply:
card-piv.c changes were removed from this PR. However, I think we should implement piv_logout correctly if possible, i.e. with explicit logout.

I don't know if disconnect_action is still needed in 2023, but if you think we should respect this when shutting down the minidriver or the PKCS#11 module, then this should be removed in [‎src/minidriver/minidriver.c‎](https://github.com/OpenSC/OpenSC/blob/6e3908fb655d3266397bfebdb328ff8f3c7206e8/src/minidriver/minidriver.c#L4091) or [‎src/pkcs11/pkcs11-session.c‎](https://github.com/OpenSC/OpenSC/blob/6e3908fb655d3266397bfebdb328ff8f3c7206e8/src/pkcs11/pkcs11-session.c#L126) respectively.